test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+        assert(accessToken.principalType);
+
+        User.findById(accessToken.userId, function(err, user) {
+          user.createAccessToken(120)
+            .then(function(accessToken) {
+              assertGoodToken(accessToken, validCredentialsUser);
+              assert.equal(accessToken.ttl, 120);
+
+              done();
+            })
+            .catch(function(err) {
+              done(err);
+            });
+        });
+      });
+    });
+
+    it('Login a user using a custom createAccessToken', function(done) {
+      var createToken = User.prototype.createAccessToken; // Save the original method
+      // Override createAccessToken
+      User.prototype.createAccessToken = function(ttl, cb) {
+        // Reduce the ttl by half for testing purpose
+        this.accessTokens.create({ttl: ttl / 2}, cb);
+      };
+      User.login(validCredentialsWithTTL, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+        assert.equal(accessToken.ttl, 1800);
+
+        User.findById(accessToken.userId, function(err, user) {
+          user.createAccessToken(120, function(err, accessToken) {
+            assertGoodToken(accessToken, validCredentialsUser);
+            assert.equal(accessToken.ttl, 60);
+            // Restore create access token
+            User.prototype.createAccessToken = createToken;
+
+            done();
+          });
+        });
+      });
+    });
+
+    it('Login a user using a custom createAccessToken with options',
+    function(done) {
+      var createToken = User.prototype.createAccessToken; // Save the original method
+      // Override createAccessToken
+      User.prototype.createAccessToken = function(ttl, options, cb) {
+        // Reduce the ttl by half for testing purpose
+        this.accessTokens.create({ttl: ttl / 2, scopes: options.scope}, cb);
+      };
+      User.login(validCredentialsWithTTLAndScope, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+        assert.equal(accessToken.ttl, 1800);
+        assert.equal(accessToken.scopes, 'all');
+
+        User.findById(accessToken.userId, function(err, user) {
+          user.createAccessToken(120, {scope: 'default'}, function(err, accessToken) {
+            assertGoodToken(accessToken, validCredentialsUser);
+            assert.equal(accessToken.ttl, 60);
+            assert.equal(accessToken.scopes, 'default');
+            // Restore create access token
+            User.prototype.createAccessToken = createToken;
+
+            done();
+          });
+        });
+      });
+    });
+
+    it('Login should only allow correct credentials', function(done) {
+      User.login(invalidCredentials, function(err, accessToken) {
+        assert(err);
+        assert.equal(err.code, 'LOGIN_FAILED');
+        assert(!accessToken);
+
+        done();
+      });
+    });
+
+    it('Login should only allow correct credentials - promise variant', function(done) {
```

Comment:
No need to test it here. Pick one variant only, the other is already covered by `test/user.test.js`.

In general, we would like to use Promise API in newly written tests. So I'd say drop the callback variant of this test and use Promise one only.